### PR TITLE
fix: [DynamoDB] Ensure that inflight persistence IDs are not evicted

### PR DESCRIPTION
**Pull request by leviramsey**
If, when recovering a backlog in an at-least-once projection, the range of event timestamps from a slice in a given batch of records to have offsets saved exceeds the offset store's `time-window`, the older events will be immediately evicted; since the inflight records are cleaned based on the state after evicting, this implies that the records for those older events will continue to be considered inflight.

If there are no further events for a persistence ID that was immediately evicted, this means a leak in inflight (if there are enough of these situations, the hard limit of 10k records inflight will get breached) eventually resulting in projection failure.

Note that sequence numbers are saved based on the incoming records, so this eviction will not prevent seen sequence numbers from being updated.

### CODE REVIEW

File: akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/DynamoDBOffsetStore.scala [hunk 1]
```diff
@@ -45,25 +45,10 @@ private[projection] object DynamoDBOffsetStore {
...
-        else Integer.compare(this.slice, that.slice)
-      } else {
-        result
-      }
-    }
-  }
+  final case class Record(slice: Int, pid: Pid, seqNr: SeqNr, timestamp: Instant)
+
+  // Used only in eviction, does not imply anything about equals
+  private val recordOrdering: Ordering[Record] = Ordering.by(_.timestamp)
```
Comment (leviramsey) [inline akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/DynamoDBOffsetStore.scala:51]: `Ordered` carries with it object equality connotations (that `a.compare(b) == 0` implies `a.equals(b)`) that we don't really need.

File: akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/DynamoDBOffsetStore.scala [hunk 2]
```diff
@@ -399,13 +408,15 @@ private[projection] class DynamoDBOffsetStore(
...
-            .toVector
+          val grouped = records.groupBy(_.pid)
+
+          if (grouped.size == records.size)
+            records.filterNot(oldState.isDuplicate)
+          else
+            grouped.valuesIterator
+              .map(_.maxBy(_.seqNr))
+              .filterNot(oldState.isDuplicate)
+              .toVector
```
Comment (patriknw) [inline akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/DynamoDBOffsetStore.scala:419]: Why is this change needed? They are already by seqNr for each pid, and that order is preserved by `groupBy(_.pid)`. Similar was in https://github.com/akka/akka-projection/pull/1260 and then reverted https://github.com/akka/akka-projection/pull/1260/commits/04c6ce46adfc3b5aeec90bfbd9cd5aee1ae2fe7b

**Review by patriknw — COMMENTED** [review]
lgtm, makes sense that we shouldn't evict those that are in fligt

but there is a change in "use last record for each pid" that I don't see why it's needed?

**Comment by patriknw** [comment]
I have incorporated this in https://github.com/akka/akka-projection/pull/1255 for r2dbc, commit https://github.com/akka/akka-projection/pull/1255/commits/22c0c26fd0832c11605603b1529e9fa44e12abbf

File: akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/DynamoDBOffsetStore.scala (hunk 2 above)
Comment (leviramsey) [reply akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/DynamoDBOffsetStore.scala:419]: I just keep coming back to not seeing where exactly the ordering is enforced: backtracking means it's not in the query source (for instance), so perhaps just a bit of paranoia here.

File: akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/DynamoDBOffsetStore.scala [hunk 3]
```diff
@@ -147,14 +143,27 @@ private[projection] object DynamoDBOffsetStore {
...
         val until = recordsSortedByTimestamp.last.timestamp.minus(timeWindow)
-        val filtered = recordsSortedByTimestamp.dropWhile(_.timestamp.isBefore(until))
-        if (filtered.size == recordsSortedByTimestamp.size) {
+        val filtered = {
+          // Records comparing >= this record by recordOrdering will definitely be kept,
+          // Records comparing < this record by recordOrdering are subject to eviction
+          // Slice will be equal, and pid will compare lexicographically less than any valid pid
+          val untilRecord = Record(slice, "", 0, until)
+          val newerRecords = recordsSortedByTimestamp.rangeImpl(Some(untilRecord), None) // inclusive of until
+          val olderRecords = recordsSortedByTimestamp.rangeImpl(None, Some(untilRecord)) // exclusive of until
```
Comment (pvlugter) [inline akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/DynamoDBOffsetStore.scala:158]: I think clearer to call `rangeFrom`/`rangeUntil` rather than the underlying `rangeImpl`:

```suggestion
          val newerRecords = recordsSortedByTimestamp.rangeFrom(untilRecord) // inclusive of until
          val olderRecords = recordsSortedByTimestamp.rangeUntil(untilRecord) // exclusive of until
```

**Review by pvlugter — APPROVED** [review]
LGTM

**Comment by leviramsey** [comment]
Refs: https://github.com/akka/akka-persistence-dynamodb/issues/95
